fontforge/featurefile.c

```diff
@@ -4736,8 +4737,10 @@ static void fea_ParseSubstitute(struct parseState *tok) {
 	} else if ( cnt>=1 && tok->type==tk_by ) {
```

Comment:
I am unfortunately now certain that this is the problem. It assumes that the _target_ of the replacement will never be non-existent, only the _replacement_. That might make sense, but I feel it will confuse users. So, a fix for this will need to be found.

Reply:
OK.

Reply:
Been debugging this and the hard error seems to be this: `This lookup has no effect, I can't figure out its type on line 9 of  test-feature.fea` which is the closing tag of `liga1`. The check on this code part is ignored by the newly introduced flag.

Will check further.

Reply:
OK, that error on `Expected ';' at statement end on line 11` is the actual hard error and is triggered from `fea_end_statement` falling through in `fea_ParseSubstitute` due to `rpl`  being `NULL` (non-existing glyph name `xdde`) and the next glyph name is `a` while `fea_end_statement` expects `;`. Back to code reading.

Reply:
For the record, I’m still working on this; but the problem space is expanded now to gracefully ignore invalid rules and keep only valid ones; covering substitutions, contextual lookups, mark positionings etc. The code base is designed to stop at error, so I think many functions need to be changed to handle the invalid rules gracefully.
For instance, if all the rules in a lookup are invalid, empty lookup is still created, which is undesirable. I’m trying to understand the code and handle as many cases as possible.

Reply:
Right...as I understood it, ignoring invalid rules was always the problem. Ignoring just some invalid rules doesn't really make most sense (especially if the flag's name is literally `ignore_invalid`) and will definitely lead to confused users.

If changing the existing code is very difficult, I would not oppose doing a first pass instead to just kick out rules with glyph names that don't exist. But that might be even harder than editing the existing code, as you'll essentially need to rewrite the parser. There are quite a few cases yes, but basically it's just a matter of, upon discovering a glyph name doesn't exist, consuming more tokens until you get to the next line.

Reply:
Hi @rajeeshknambiar , everything OK? Did I miss something about how difficult this feature is? Do you maybe want to collaborate a little more to get it merged? I already wrote some code for you, and I think this feature is genuinely useful, so I wouldn't mind helping out a little more in the code department.

Reply:
Hey Frederick,

Thanks for the kind offer. I’ve been quite busy with work and couldn’t dedicate as much time I wanted to on the feature parsing changes. The rough plan I have had is:

1.  Check all the different functions called from `fea_LookupSwitch` which then fills the `feat_item *item` linked list.
2.  Right now the code seems to fill `item` list with tokens as and when it is parsed correctly.
3.  Behaviour in step 2 may need to be changed to fill a temporary list by each function (for parsing one whole lookup rule ) and if it is valid, add it to the `item` list.

I haven’t got to wrapping my head around all these. How would you like to proceed with collaborating? I’m all ears.